### PR TITLE
refactoring of packet io

### PR DESCRIPTION
**Pull request by toanju**
* remove tap_manager from cbasebox
* add enqueue functions to sai and implement them accordingly, since
  tap_callback has to be removed
* new ports have to be notified via sai to be created

### CODE REVIEW

File: src/roflibs/of-dpa/cbasebox.cpp [hunk 1]
```diff
@@ -391,37 +379,35 @@ int cbasebox::enqueue(rofcore::ctapdev *tapdev, rofl::cpacket *pkt) {
...
       rv = -ENOTCONN;
       goto errout;
     }
-
-    // TODO move to separate function:
-    uint32_t portno =
-        dpt.get_ports().get_port(tapdev->get_devname()).get_port_no();
-
+    
     /* only send packet-out if we can determine a port-no */
```
Comment (DFritzsche) [inline src/roflibs/of-dpa/cbasebox.cpp]: port-no is not used anymore in the code so the comment should be updated as well

File: src/roflibs/of-dpa/cbasebox.cpp [hunk 2]
```diff
@@ -208,7 +207,25 @@ void cbasebox::handle_port_desc_stats_reply(
...
+        nbi::PORT_EVENT_ADD, port.get_port_no(), port.get_name()});
+  }
+
+  /* init 1:1 port mapping */
+  try {
+    this->nbi->port_notification(notifications);
+    LOG(INFO) << "ports initialized";
+
+  } catch (std::exception &e) {
+    LOG(ERROR) << __FUNCTION__ << "] ERROR: unknown error " << e.what();
```
Comment (DFritzsche) [inline src/roflibs/of-dpa/cbasebox.cpp]: in some error msgs, you deleted the "]", this should be consistent

File: src/roflibs/netlink/tap_manager.hpp [hunk 3]
```diff
@@ -47,10 +50,10 @@ class tap_manager final {
   tap_manager(const tap_manager &other) = delete; // non construction-copyable
   tap_manager &operator=(const tap_manager &) = delete; // non copyable
 
-  int create_tapdev(const std::string &, tap_callback &);
+  int create_tapdev(uint32_t, const std::string &, tap_callback &);
```
Comment (DFritzsche) [inline src/roflibs/netlink/tap_manager.hpp]: I do not like the fact, that sometimes you leave out parameter names in function declarations

File: src/baseboxd/baseboxd.cpp [hunk 4]
```diff
@@ -33,9 +35,7 @@ int main(int argc, char **argv) {
   versionbitmap.add_ofp_version(rofl::openflow13::OFP_VERSION);
   LOG(INFO) << "using OpenFlow version-bitmap:" << std::endl << versionbitmap;
 
-  // start netlink
-  rofcore::nbi *nbi = &rofcore::cnetlink::get_instance();
-  // XXX FIXME move tap_manager here
+  rofcore::nbi_impl *nbi = new rofcore::nbi_impl();
```
Comment (DFritzsche) [inline src/baseboxd/baseboxd.cpp]: I can not find the `delete` for this object

File: src/roflibs/netlink/tap_manager.hpp (hunk 3 above)
Comment (toanju) [inline src/roflibs/netlink/tap_manager.hpp]: we should agree on a general sytle here :)

File: src/roflibs/of-dpa/cbasebox.cpp (hunk 2 above)
Comment (toanju) [inline src/roflibs/of-dpa/cbasebox.cpp]: ack

File: src/roflibs/of-dpa/cbasebox.cpp (hunk 1 above)
Comment (toanju) [inline src/roflibs/of-dpa/cbasebox.cpp]: ups will remove this

File: src/baseboxd/baseboxd.cpp (hunk 4 above)
Comment (toanju) [inline src/baseboxd/baseboxd.cpp]: missing

Comment (toanju) [inline src/baseboxd/baseboxd.cpp]: added and all pushed

**Pull request merged**